homeassistant/components/goalzero/config_flow.py

```diff
@@ -19,51 +24,99 @@ class GoalZeroFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+                CONF_NAME: DEFAULT_NAME,
+            },
+        )
 
     async def async_step_user(self, user_input=None):
         """Handle a flow initiated by the user."""
-        errors = {}
-
         if user_input is not None:
             host = user_input[CONF_HOST]
             name = user_input[CONF_NAME]
 
-            self._async_abort_entries_match({CONF_HOST: host})
-
-            try:
-                await self._async_try_connect(host)
-            except exceptions.ConnectError:
-                errors["base"] = "cannot_connect"
-                _LOGGER.error("Error connecting to device at %s", host)
-            except exceptions.InvalidHost:
-                errors["base"] = "invalid_host"
-                _LOGGER.error("Invalid host at %s", host)
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Unexpected exception")
-                errors["base"] = "unknown"
-            else:
-                return self.async_create_entry(
-                    title=name,
-                    data={CONF_HOST: host, CONF_NAME: name},
-                )
+            await self.async_try_connect(host, name)
 
         user_input = user_input or {}
         return self.async_show_form(
             step_id="user",
             data_schema=vol.Schema(
                 {
-                    vol.Required(
-                        CONF_HOST, default=user_input.get(CONF_HOST) or ""
-                    ): str,
+                    vol.Required(CONF_HOST, default=self.ip_address): str,
                     vol.Optional(
                         CONF_NAME, default=user_input.get(CONF_NAME) or DEFAULT_NAME
                     ): str,
                 }
             ),
-            errors=errors,
+            errors=self.errors,
         )
 
-    async def _async_try_connect(self, host):
-        session = async_get_clientsession(self.hass)
-        api = Yeti(host, self.hass.loop, session)
-        await api.get_state()
+    async def async_try_connect(self, host, name):
+        """Try connecting."""
+        try:
+            session = async_get_clientsession(self.hass)
+            api = Yeti(host, self.hass.loop, session)
+            await api.init_connect()
+            await self.async_set_unique_id(api.sysdata["macAddress"].lower())
+            self._abort_if_unique_id_configured(updates={CONF_HOST: host})
+            for entry in self._async_current_entries():
+                if entry.data.get(CONF_HOST) == host:
+                    return self.async_abort(reason="already_configured")
+                CONF_DATA = {
+                    CONF_HOST: host,
+                    CONF_NAME: name,
+                }
+                self.hass.config_entries.async_update_entry(entry, data=CONF_DATA)
+                await self.hass.config_entries.async_reload(entry.entry_id)
+        except exceptions.ConnectError:
+            self.errors["base"] = "cannot_connect"
```

Comment:
Please don't share errors between steps. The errors are specific and private to the form of the step that generated them. We can raise an exception instead.